File: src/main/java/com/itszuvalex/femtocraft/player/PlayerProperties.java
/*
 * ******************************************************************************
 *  * Copyright (C) 2013  Christopher Harris (Itszuvalex)
 *  * Itszuvalex@gmail.com
 *  *
 *  * This program is free software; you can redistribute it and/or
 *  * modify it under the terms of the GNU General Public License
 *  * as published by the Free Software Foundation; either version 2
 *  * of the License, or (at your option) any later version.
 *  *
 *  * This program is distributed in the hope that it will be useful,
 *  * but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  * GNU General Public License for more details.
 *  *
 *  * You should have received a copy of the GNU General Public License
 *  * along with this program; if not, write to the Free Software
 *  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *  *****************************************************************************
 */

package com.itszuvalex.femtocraft.player;

import com.itszuvalex.femtocraft.Femtocraft;
import com.itszuvalex.femtocraft.network.FemtocraftPacketHandler;
import com.itszuvalex.femtocraft.network.messages.MessagePlayerProperty;
import cpw.mods.fml.common.FMLCommonHandler;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.world.World;
import net.minecraftforge.common.IExtendedEntityProperties;

import java.util.Map;
import java.util.TreeMap;
import java.util.logging.Level;

public class PlayerProperties implements IExtendedEntityProperties {
    public final static String PROP_TAG = "femtocraft.player.properties";
    public final static String PACKET_CHANNEL = Femtocraft.PLAYER_PROP_CHANNEL();

    protected final EntityPlayer player;

    private Map<String, IFemtocraftPlayerProperty> properties = new TreeMap<String, IFemtocraftPlayerProperty>();

    private static Map<String, Class<? extends IFemtocraftPlayerProperty>> propertiesClasses = new TreeMap<String,
            Class<?
                    extends IFemtocraftPlayerProperty>>();

    public static boolean registerPlayerProperty(String name, Class<? extends IFemtocraftPlayerProperty>
            propertyClass) {
        if (propertiesClasses.containsKey(name)) return false;
        return propertiesClasses.put(name, propertyClass) != null;
    }

    public IFemtocraftPlayerProperty getProperty(String name) {
        return properties.get(name);
    }

    public PlayerProperties(EntityPlayer player) {
        this.player = player;
        for (Map.Entry<String, Class<? extends IFemtocraftPlayerProperty>> entry : propertiesClasses
                .entrySet()) {

            try {
                properties.put(entry.getKey(), entry.getValue().newInstance());
            } catch (InstantiationException e) {
                Femtocraft.logger().log(Level.SEVERE, "Failed to create new instance of " + entry.getKey() +
                                                      " on creating PlayerProperties for player: " + player +
                                                      " name: " +
                                                      player.getCommandSenderName());
                e.printStackTrace();
            } catch (IllegalAccessException e) {
                e.printStackTrace();
            }
        }
    }

    public static void register(EntityPlayer player) {
        player.registerExtendedProperties(PROP_TAG,
                new PlayerProperties(player));
    }

    public static PlayerProperties get(EntityPlayer player) {
        return (PlayerProperties) player.getExtendedProperties(PROP_TAG);
    }

    public void sync() {
        if (FMLCommonHandler.instance().getEffectiveSide().isClient()) {
            return;
        }

        NBTTagCompound compound = new NBTTagCompound();

        for (Map.Entry<String, IFemtocraftPlayerProperty> entry : properties.entrySet()) {
            savePropertyToCompound(entry.getKey(), compound);
        }

        FemtocraftPacketHandler.INSTANCE().sendTo(new MessagePlayerProperty(player.getCommandSenderName(), compound),
                (EntityPlayerMP) player);
    }

    public void sync(String property) {
        if (FMLCommonHandler.instance().getEffectiveSide().isClient()) return;

        NBTTagCompound packetCompound = new NBTTagCompound();
        savePropertyToCompound(property, packetCompound);
        FemtocraftPacketHandler.INSTANCE().sendTo(new MessagePlayerProperty(player.getCommandSenderName(),
                packetCompound), (EntityPlayerMP) player);
    }

    private void savePropertyToCompound(String property, NBTTagCompound packetCompound) {
        if (FMLCommonHandler.instance().getEffectiveSide().isClient()) {
            return;
        }

        IFemtocraftPlayerProperty prop = properties.get(property);
        if (prop != null) {
            NBTTagCompound propCompound = new NBTTagCompound();
            prop.toDescriptionPacket(propCompound);
            packetCompound.setTag(property, propCompound);

        }
    }

    @Override
    public void saveNBTData(NBTTagCompound compound) {
        NBTTagCompound properties = new NBTTagCompound();
        for (Map.Entry<String, IFemtocraftPlayerProperty> entry : this.properties.entrySet()) {
            NBTTagCompound entryComp = new NBTTagCompound();
            entry.getValue().saveToNBT(entryComp);
            properties.setTag(entry.getKey(), entryComp);
        }

        compound.setTag(PROP_TAG, properties);
    }

    @Override
    public void loadNBTData(NBTTagCompound compound) {
        NBTTagCompound properties = compound.getCompoundTag(PROP_TAG);
        for (Map.Entry<String, IFemtocraftPlayerProperty> entry : this.properties.entrySet()) {
            entry.getValue().loadFromNBT(properties.getCompoundTag(entry.getKey()));
        }
    }

    @Override
    public void init(Entity entity, World world) {
    }

    public void handlePacket(NBTTagCompound compound) {
        for (Map.Entry<String, IFemtocraftPlayerProperty> entry : this.properties.entrySet()) {
            if (compound.hasKey(entry.getKey())) {
                entry.getValue().loadFromDescription(compound.getCompoundTag(entry.getKey()));
            }
        }
    }

}

File: src/main/java/com/itszuvalex/femtocraft/player/PlayerProperties.scala
/*
 * ******************************************************************************
 *  * Copyright (C) 2013  Christopher Harris (Itszuvalex)
 *  * Itszuvalex@gmail.com
 *  *
 *  * This program is free software; you can redistribute it and/or
 *  * modify it under the terms of the GNU General Public License
 *  * as published by the Free Software Foundation; either version 2
 *  * of the License, or (at your option) any later version.
 *  *
 *  * This program is distributed in the hope that it will be useful,
 *  * but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  * GNU General Public License for more details.
 *  *
 *  * You should have received a copy of the GNU General Public License
 *  * along with this program; if not, write to the Free Software
 *  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *  *****************************************************************************
 */
package com.itszuvalex.femtocraft.player

import java.util

import com.itszuvalex.femtocraft.Femtocraft
import com.itszuvalex.femtocraft.api.player.IFemtocraftPlayerProperty
import com.itszuvalex.femtocraft.network.FemtocraftPacketHandler
import com.itszuvalex.femtocraft.network.messages.MessagePlayerProperty
import cpw.mods.fml.common.FMLCommonHandler
import net.minecraft.entity.Entity
import net.minecraft.entity.player.{EntityPlayer, EntityPlayerMP}
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.world.World
import net.minecraftforge.common.IExtendedEntityProperties
import org.apache.logging.log4j.Level

import scala.collection.JavaConversions._

object PlayerProperties {
  val PROP_TAG = "femtocraft.player.properties"
  private val propertiesClasses = new util.TreeMap[String, Class[_ <: IFemtocraftPlayerProperty]]

  def registerPlayerProperty(name: String, propertyClass: Class[_ <: IFemtocraftPlayerProperty]): Boolean = {
    if (propertiesClasses.containsKey(name)) return false
    propertiesClasses.put(name, propertyClass) != null
  }

  def register(player: EntityPlayer) = player.registerExtendedProperties(PROP_TAG, new PlayerProperties(player))

  def get(player: EntityPlayer) = player.getExtendedProperties(PROP_TAG).asInstanceOf[PlayerProperties]
}

class PlayerProperties(protected final var player: EntityPlayer) extends IExtendedEntityProperties {
  private val properties = new util.TreeMap[String, IFemtocraftPlayerProperty]


  for (entry <- PlayerProperties.propertiesClasses.entrySet) {
    try {
      properties.put(entry.getKey, entry.getValue.newInstance)
    }
    catch {
      case e: InstantiationException =>
        Femtocraft.log(Level.ERROR, "Failed to create new instance of " + entry.getKey + " on creating PlayerProperties for player: " + player + " name: " + player.getCommandSenderName)
        e.printStackTrace()
      case e: IllegalAccessException =>
        e.printStackTrace()
    }
  }

  def getProperty(name: String) = properties.get(name)

  def this() = this(null)

  def sync() {
    if (FMLCommonHandler.instance.getEffectiveSide.isClient) {
      return
    }
    val compound = new NBTTagCompound

    for (entry <- properties.entrySet) {
      savePropertyToCompound(entry.getKey, compound)
    }
    FemtocraftPacketHandler.INSTANCE.sendTo(new MessagePlayerProperty(player.getCommandSenderName, compound), player.asInstanceOf[EntityPlayerMP])
  }

  def sync(property: String) {
    if (FMLCommonHandler.instance.getEffectiveSide.isClient) return
    val packetCompound = new NBTTagCompound
    savePropertyToCompound(property, packetCompound)
    FemtocraftPacketHandler.INSTANCE.sendTo(new MessagePlayerProperty(player.getCommandSenderName, packetCompound), player.asInstanceOf[EntityPlayerMP])
  }

  private def savePropertyToCompound(property: String, packetCompound: NBTTagCompound) {
    if (FMLCommonHandler.instance.getEffectiveSide.isClient) return
    val prop = properties.get(property)
    if (prop != null) {
      val propCompound = new NBTTagCompound
      prop.toDescriptionPacket(propCompound)
      packetCompound.setTag(property, propCompound)
    }
  }

  def saveNBTData(compound: NBTTagCompound) {
    val propTag = new NBTTagCompound
    for (entry <- properties.entrySet) {
      val entryComp = new NBTTagCompound
      entry.getValue.saveToNBT(entryComp)
      propTag.setTag(entry.getKey, entryComp)
    }
    compound.setTag(PlayerProperties.PROP_TAG, propTag)
  }

  def loadNBTData(compound: NBTTagCompound) {
    val propTag = compound.getCompoundTag(PlayerProperties.PROP_TAG)
    for (entry <- properties.entrySet) {
      entry.getValue.loadFromNBT(propTag.getCompoundTag(entry.getKey))
    }
  }

  def init(entity: Entity, world: World) {
  }

  def handlePacket(compound: NBTTagCompound) {
    for (entry <- properties.entrySet) {
      if (compound.hasKey(entry.getKey)) {
        entry.getValue.loadFromDescription(compound.getCompoundTag(entry.getKey))
      }
    }
  }
}
